### PR TITLE
Use HiveCatalogName to export S3 statistics without name conflicts

### PR DESCRIPTION
**Pull request by MiguelWeezardo**
This fixes the problem of conflicting JMX keys for S3 statistics
when Hive and Iceberg connectors are loaded.

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/InternalHiveConnectorFactory.java [hunk 1]
```diff
@@ -81,7 +81,7 @@ public static Connector createConnector(String catalogName, Map<String, String>
                     new ConnectorObjectNameGeneratorModule(catalogName, "io.prestosql.plugin.hive", "presto.plugin.hive"),
                     new JsonModule(),
                     new HiveModule(),
-                    new HiveS3Module(),
+                    new HiveS3Module(catalogName),
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/InternalHiveConnectorFactory.java]: There is HiveCatalogName bean binded already

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Module.java [hunk 2]
```diff
@@ -39,7 +47,7 @@ protected void setup(Binder binder)
 
             binder.bind(PrestoS3FileSystemStats.class).toInstance(PrestoS3FileSystem.getFileSystemStats());
             newExporter(binder).export(PrestoS3FileSystemStats.class)
-                    .as(generator -> generator.generatedNameOf(PrestoS3FileSystem.class));
+                    .as(generator -> generator.generatedNameOf(PrestoS3FileSystem.class, catalogName));
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Module.java]: You can use provider of HiveCatalogName and pass function here to generate name lazy. After guice configuration

Comment (MiguelWeezardo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Module.java]: I tried to replace `catalogName` with `binder.getProvider(HiveCatalogName.class).get().toString()` but I'm getting the following error:
```
2020-03-13T16:45:14.015+0100	ERROR	main	io.prestosql.server.PrestoServer	Unable to create injector, see the following errors:

1) Error injecting constructor, java.lang.NullPointerException: moduleSource cannot be null.
  at org.weakref.jmx.guice.GuiceMBeanExporter.<init>(GuiceMBeanExporter.java:40)
  at org.weakref.jmx.guice.MBeanModule.configure(MBeanModule.java:39)
  while locating org.weakref.jmx.guice.GuiceMBeanExporter

1 error
com.google.inject.CreationException: Unable to create injector, see the following errors:

1) Error injecting constructor, java.lang.NullPointerException: moduleSource cannot be null.
  at org.weakref.jmx.guice.GuiceMBeanExporter.<init>(GuiceMBeanExporter.java:40)
  at org.weakref.jmx.guice.MBeanModule.configure(MBeanModule.java:39)
  while locating org.weakref.jmx.guice.GuiceMBeanExporter

1 error
	at com.google.inject.internal.Errors.throwCreationExceptionIfErrorsExist(Errors.java:543)
	at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:186)
	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:109)
	at com.google.inject.Guice.createInjector(Guice.java:87)
	at io.airlift.bootstrap.Bootstrap.initialize(Bootstrap.java:245)
	at io.prestosql.plugin.hive.InternalHiveConnectorFactory.createConnector(InternalHiveConnectorFactory.java:96)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.prestosql.plugin.hive.HiveConnectorFactory.create(HiveConnectorFactory.java:52)
	at io.prestosql.connector.ConnectorManager.createConnector(ConnectorManager.java:341)
	at io.prestosql.connector.ConnectorManager.createCatalog(ConnectorManager.java:203)
	at io.prestosql.connector.ConnectorManager.createCatalog(ConnectorManager.java:195)
	at io.prestosql.connector.ConnectorManager.createCatalog(ConnectorManager.java:181)
	at io.prestosql.metadata.StaticCatalogStore.loadCatalog(StaticCatalogStore.java:88)
	at io.prestosql.metadata.StaticCatalogStore.loadCatalogs(StaticCatalogStore.java:68)
	at io.prestosql.server.PrestoServer.run(PrestoServer.java:129)
	at io.prestosql.$gen.Presto_null____20200313_154403_1.run(Unknown Source)
	at io.prestosql.server.PrestoServer.main(PrestoServer.java:72)
Caused by: java.lang.NullPointerException: moduleSource cannot be null.
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:895)
	at com.google.inject.spi.ElementSource.<init>(ElementSource.java:94)
	at com.google.inject.spi.Elements$RecordingBinder.getElementSource(Elements.java:561)
	at com.google.inject.spi.Elements$RecordingBinder.getProvider(Elements.java:411)
	at com.google.inject.spi.Elements$RecordingBinder.getProvider(Elements.java:406)
	at com.google.inject.spi.Elements$RecordingBinder.getProvider(Elements.java:418)
	at io.prestosql.plugin.hive.s3.HiveS3Module.lambda$setup$0(HiveS3Module.java:43)
	at org.weakref.jmx.guice.Mapping.getName(Mapping.java:36)
	at org.weakref.jmx.guice.GuiceMBeanExporter.export(GuiceMBeanExporter.java:92)
	at org.weakref.jmx.guice.GuiceMBeanExporter.<init>(GuiceMBeanExporter.java:42)
	at org.weakref.jmx.guice.GuiceMBeanExporter$$FastClassByGuice$$6cc63b28.newInstance(<generated>)
	at com.google.inject.internal.DefaultConstructionProxyFactory$FastClassProxy.newInstance(DefaultConstructionProxyFactory.java:89)
	at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:114)
	at com.google.inject.internal.ConstructorInjector.access$000(ConstructorInjector.java:32)
	at com.google.inject.internal.ConstructorInjector$1.call(ConstructorInjector.java:98)
	at com.google.inject.internal.ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:112)
	at io.airlift.bootstrap.LifeCycleModule.provision(LifeCycleModule.java:54)
	at com.google.inject.internal.ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:120)
	at com.google.inject.internal.ProvisionListenerStackCallback.provision(ProvisionListenerStackCallback.java:66)
	at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:93)
	at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:306)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
	at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:168)
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:39)
	at com.google.inject.internal.InternalInjectorCreator.loadEagerSingletons(InternalInjectorCreator.java:211)
	at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:182)
	... 18 more
```
I'm open to suggestions, it seems the binder isn't ready to create a provider at this point (we're still during module setup).

**Comment by findepi** [comment]
CI failed -- https://github.com/prestosql/presto/issues/3049

**Comment by findepi** [comment]
@MiguelWeezardo please force push to rerun CI.

**Comment by findepi** [comment]
@kokosing i liked the previous version (https://github.com/prestosql/presto/commit/325a8c6c59d0a452b5f31247c66cc689cf694215) much better.
Why not KISS?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Module.java [hunk 3]
```diff
@@ -38,8 +41,9 @@ protected void setup(Binder binder)
             configBinder(binder).bindConfig(HiveS3Config.class);
 
             binder.bind(PrestoS3FileSystemStats.class).toInstance(PrestoS3FileSystem.getFileSystemStats());
+            Provider<HiveCatalogName> catalogNameProvider = binder.getProvider(HiveCatalogName.class);
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Module.java]: just `catalogName`?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Module.java [hunk 4]
```diff
@@ -53,6 +57,11 @@ else if (type == S3FileSystemType.HADOOP_DEFAULT) {
         }
     }
 
+    private String generateStatisticNameForCatalog(Class<?> type, ObjectNameGenerator generator, Provider<HiveCatalogName> catalogNameProvider)
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Module.java]: I would inline this method.

**Comment by kokosing** [comment]
> Why not KISS?

I don't find this approach much harder than the previous one. Maybe it just require to know more details about guice. I see following issues with previous approach:
 - it requires to pass variable through constructors, it makes it a bit more difficult to maintain module classes. For example when you move this module you might need to update constructors.
 - it is opposite to guice
 - it is opposite to `HiveCatalogName` bean purpose. If we are not able to use it to access catalog name in guive then it means that `HiveCatalogName` is somehow broken. 
 - this approach works well work for additional modules, so same thing is implemented in the same way across different modules.

**Comment by MiguelWeezardo** [comment]
@kokosing @findepi If tests pass, could one of you merge this? I don't have permissions for that.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Module.java [hunk 5]
```diff
@@ -38,8 +40,9 @@ protected void setup(Binder binder)
             configBinder(binder).bindConfig(HiveS3Config.class);
 
             binder.bind(PrestoS3FileSystemStats.class).toInstance(PrestoS3FileSystem.getFileSystemStats());
+            Provider<HiveCatalogName> catalogName = binder.getProvider(HiveCatalogName.class);
             newExporter(binder).export(PrestoS3FileSystemStats.class)
-                    .as(generator -> generator.generatedNameOf(PrestoS3FileSystem.class));
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Module.java]: formatting change is unrelated and incorrect

**Comment by MiguelWeezardo** [comment]
@kokosing @findepi Tests failed twice with the same error:
`com.github.dockerjava.api.exception.DockerClientException: Could not pull image: failed to register layer: Error processing tar file(exit status 1): write /usr/hdp/3.1.0.0-78/hive/hive.tar.gz: no space left on device`
Is there anything I can do about this? Or should I just keep changing the commit timestamps and hope for the best?

**Comment by electrum** [comment]
Per https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#supported-runners-and-hardware-resources

> 14 GB of SSD disk space

I wonder if we were always close, or if something changed recently.

**Comment by kokosing** [comment]
Merged, thanks!
